x/ccv/provider/types/msg.go

```diff
@@ -203,3 +209,163 @@ func (msg MsgSubmitConsumerDoubleVoting) GetSigners() []sdk.AccAddress {
+
+// GetSigners implements the sdk.Msg interface. It returns the address(es) that
+// must sign over msg.GetSignBytes().
+func (msg MsgOptIn) GetSigners() []sdk.AccAddress {
+	valAddr, err := sdk.ValAddressFromBech32(msg.ProviderAddr)
+	if err != nil {
+		// same behavior as in cosmos-sdk
+		panic(err)
+	}
+	return []sdk.AccAddress{valAddr.Bytes()}
+}
+
+// GetSignBytes returns the message bytes to sign over.
+func (msg MsgOptIn) GetSignBytes() []byte {
+	bz := ccvtypes.ModuleCdc.MustMarshalJSON(&msg)
+	return sdk.MustSortJSON(bz)
+}
+
+// ValidateBasic implements the sdk.Msg interface.
+func (msg MsgOptIn) ValidateBasic() error {
+	if strings.TrimSpace(msg.ChainId) == "" {
+		return errorsmod.Wrapf(ErrInvalidConsumerChainID, "chainId cannot be blank")
+	}
+	// It is possible to assign keys for consumer chains that are not yet approved.
+	// This can only be done by a signing validator, but it is still sensible
+	// to limit the chainID size to prevent abuse.
+	if 128 < len(msg.ChainId) {
+		return errorsmod.Wrapf(ErrInvalidConsumerChainID, "chainId cannot exceed 128 length")
+	}
+	_, err := sdk.ValAddressFromBech32(msg.ProviderAddr)
+	if err != nil {
+		return ErrInvalidProviderAddress
+	}
+
+	if msg.ConsumerKey != "" {
+		if _, _, err := ParseConsumerKeyFromJson(msg.ConsumerKey); err != nil {
+			return ErrInvalidConsumerConsensusPubKey
+		}
+	}
+	return nil
+}
+
+// NewMsgOptOut creates a new NewMsgOptIn instance.
+func NewMsgOptOut(chainID string, providerValidatorAddress sdk.ValAddress) (*MsgOptOut, error) {
+	return &MsgOptOut{
+		ChainId:      chainID,
+		ProviderAddr: providerValidatorAddress.String(),
+	}, nil
+}
+
+// Route implements the sdk.Msg interface.
+func (msg MsgOptOut) Route() string { return RouterKey }
+
+// Type implements the sdk.Msg interface.
+func (msg MsgOptIn) Type() string {
+	return TypeMsgOptIn
+}
+
+// GetSigners implements the sdk.Msg interface. It returns the address(es) that
+// must sign over msg.GetSignBytes().
+func (msg MsgOptOut) GetSigners() []sdk.AccAddress {
+	valAddr, err := sdk.ValAddressFromBech32(msg.ProviderAddr)
+	if err != nil {
+		// same behavior as in cosmos-sdk
+		panic(err)
+	}
+	return []sdk.AccAddress{valAddr.Bytes()}
+}
+
+// GetSignBytes returns the message bytes to sign over.
+func (msg MsgOptOut) GetSignBytes() []byte {
+	bz := ccvtypes.ModuleCdc.MustMarshalJSON(&msg)
+	return sdk.MustSortJSON(bz)
+}
+
+// ValidateBasic implements the sdk.Msg interface.
+func (msg MsgOptOut) ValidateBasic() error {
+	if strings.TrimSpace(msg.ChainId) == "" {
+		return errorsmod.Wrapf(ErrInvalidConsumerChainID, "chainId cannot be blank")
+	}
+	// It is possible to assign keys for consumer chains that are not yet approved.
+	// This can only be done by a signing validator, but it is still sensible
+	// to limit the chainID size to prevent abuse.
+	if 128 < len(msg.ChainId) {
+		return errorsmod.Wrapf(ErrInvalidConsumerChainID, "chainId cannot exceed 128 length")
+	}
```

Comment:
Not needed as we return error when handling the message if the chain ID is not found. There is not risk to spam the store. 
```suggestion
```

Reply:
See above